# DataCo Smart Supply Chain — exploration + classical baseline

External real-world dataset (separate from the synthetic `scm_dataset` benchmark) used as a second track: reproduce a legitimate classical baseline here, then later build a QGNN on the same graph/split for a direct comparison.

**Source**: Constante, Silva & Pereira (2019), Mendeley Data V5, DOI `10.17632/8gx2fvg2k6.5`, CC-BY-4.0. Downloaded via `scripts/download_dataco.py` -> `data/raw/dataco/`. Provenance: `data/raw/dataco/provenance.json`.

**Why not reproduce the "Integrated Machine Learning for Enhanced Supply Chain Risk Prediction" preprint (Jin, 2025, Preprints.org, not peer-reviewed)**: its Figure 4 lists features ("Lead Time", "Demand Variation", "Supplier Reliability", "Inventory Level", "Order Cost") that **do not exist** in the actual public DataCo schema (verified below) despite the paper claiming to use DataCo Global's data. That mismatch, combined with the "not peer-reviewed" status, means its reported 85.4% accuracy / 0.85 F1 isn't a trustworthy reproduction target. This notebook builds an honest baseline from the real columns instead, and treats peer-reviewed / independently-reproducible numbers (e.g. the EAGLE paper's F1=0.876 graph-based result, or a plain from-scratch Random Forest) as the actual comparison points.

In [1]:
import pandas as pd
import numpy as np

DATA_PATH = "../data/raw/dataco/DataCoSupplyChainDataset.csv"
df = pd.read_csv(DATA_PATH, encoding="latin-1")
df.shape

(180519, 53)

## 1. Schema audit

Confirm the real columns (this is what caught the preprint's fabricated feature list above) and check the target's class balance.

In [2]:
list(df.columns)

['Type',
 'Days for shipping (real)',
 'Days for shipment (scheduled)',
 'Benefit per order',
 'Sales per customer',
 'Delivery Status',
 'Late_delivery_risk',
 'Category Id',
 'Category Name',
 'Customer City',
 'Customer Country',
 'Customer Email',
 'Customer Fname',
 'Customer Id',
 'Customer Lname',
 'Customer Password',
 'Customer Segment',
 'Customer State',
 'Customer Street',
 'Customer Zipcode',
 'Department Id',
 'Department Name',
 'Latitude',
 'Longitude',
 'Market',
 'Order City',
 'Order Country',
 'Order Customer Id',
 'order date (DateOrders)',
 'Order Id',
 'Order Item Cardprod Id',
 'Order Item Discount',
 'Order Item Discount Rate',
 'Order Item Id',
 'Order Item Product Price',
 'Order Item Profit Ratio',
 'Order Item Quantity',
 'Sales',
 'Order Item Total',
 'Order Profit Per Order',
 'Order Region',
 'Order State',
 'Order Status',
 'Order Zipcode',
 'Product Card Id',
 'Product Category Id',
 'Product Description',
 'Product Image',
 'Product Name',
 'Product P

In [3]:
df["Late_delivery_risk"].value_counts(normalize=True)

Late_delivery_risk
1    0.548291
0    0.451709
Name: proportion, dtype: float64

## 2. Feature audit (leakage prevention)

Same principle as `feature_audit.csv` in the synthetic benchmark (plan §27): mark every column allowed/disallowed as a model input, with a reason, *before* touching modeling code.

**Disallowed (label / post-outcome leakage)**
- `Late_delivery_risk` -- the target itself
- `Delivery Status` -- directly encodes whether delivery was late (near-tautological with the target)
- `Days for shipping (real)` -- the *actual* transit time, only known after delivery completes
- `shipping date (DateOrders)` -- the actual ship date, also only known after the fact
- `Order Status` -- fulfillment status observed progressively after order placement (e.g. `CANCELED`, `SUSPECTED_FRAUD`), not known at prediction time

**Disallowed (PII / identifiers, never modeled on regardless of predictive value)**
- `Customer Email`, `Customer Fname`, `Customer Lname`, `Customer Password`, `Customer Street`
- `Product Description`, `Product Image` (free text / URLs, no modeling value)

**Allowed (known at order time)**
- `Type`, `Days for shipment (scheduled)`, `Category Id`, `Customer Segment`, `Customer Country`, `Customer State`, `Department Id`, `Market`, `Order Region`, `Order Country`, `Order State`, `Order Item Discount(...)`, `Order Item Product Price`, `Order Item Quantity`, `Sales`, `Product Price`, `Shipping Mode`, `order date (DateOrders)` (used for the temporal split, not as a feature itself)

In [4]:
TARGET = "Late_delivery_risk"

DISALLOWED = [
    "Delivery Status", "Days for shipping (real)", "shipping date (DateOrders)", "Order Status",
    "Customer Email", "Customer Fname", "Customer Lname", "Customer Password", "Customer Street",
    "Product Description", "Product Image",
]

CATEGORICAL_FEATURES = [
    "Type", "Category Id", "Customer Segment", "Customer Country", "Customer State",
    "Department Id", "Market", "Order Region", "Order Country", "Order State", "Shipping Mode",
]
NUMERIC_FEATURES = [
    "Days for shipment (scheduled)", "Order Item Discount", "Order Item Discount Rate",
    "Order Item Product Price", "Order Item Quantity", "Sales", "Product Price",
]

assert TARGET not in CATEGORICAL_FEATURES + NUMERIC_FEATURES
assert not set(DISALLOWED) & set(CATEGORICAL_FEATURES + NUMERIC_FEATURES)
df[["order date (DateOrders)", TARGET] + CATEGORICAL_FEATURES + NUMERIC_FEATURES].isna().sum()

order date (DateOrders)          0
Late_delivery_risk               0
Type                             0
Category Id                      0
Customer Segment                 0
Customer Country                 0
Customer State                   0
Department Id                    0
Market                           0
Order Region                     0
Order Country                    0
Order State                      0
Shipping Mode                    0
Days for shipment (scheduled)    0
Order Item Discount              0
Order Item Discount Rate         0
Order Item Product Price         0
Order Item Quantity              0
Sales                            0
Product Price                    0
dtype: int64

## 3. Temporal split

Chronological, not random-shuffled (same reasoning as the synthetic benchmark's temporal split -- a random row shuffle of ordered transactions leaks future information into training).

In [5]:
df["order date (DateOrders)"] = pd.to_datetime(df["order date (DateOrders)"])
df_sorted = df.sort_values("order date (DateOrders)").reset_index(drop=True)

n = len(df_sorted)
train_end, val_end = int(n * 0.70), int(n * 0.85)
df_sorted["split"] = "test"
df_sorted.loc[:train_end, "split"] = "train"
df_sorted.loc[train_end:val_end, "split"] = "validation"
df_sorted["split"].value_counts()

split
train         126363
validation     27079
test           27077
Name: count, dtype: int64

### 3a. Split balance diagnostic

Confirms the temporal split above isn't hiding a distribution shift -- if the positive rate or date range looks wildly different across splits, the F1/accuracy numbers in section 4 aren't directly comparable across train/val/test, and any QGNN result built on this same split inherits the same distortion (see discussion: this split is reused unchanged for the QGNN graph construction in section 5).

In [6]:
split_summary = df_sorted.groupby("split").agg(
    n=(TARGET, "size"),
    positive_rate=(TARGET, "mean"),
    date_min=("order date (DateOrders)", "min"),
    date_max=("order date (DateOrders)", "max"),
).loc[["train", "validation", "test"]]
split_summary

,n,positive_rate,date_min,date_max
split,,,,
train,126363,0.549425,2015-01-01 00:00:00,2017-01-08 02:49:00
validation,27079,0.541083,2017-01-08 02:49:00,2017-06-14 07:19:00
test,27077,0.550209,2017-06-14 07:19:00,2018-01-31 23:38:00


## 4. Preprocessing + classical baselines

Logistic Regression (matches the synthetic benchmark's non-graph baseline in `src/scm_dataset/modeling/baselines.py`) and Random Forest (the strongest single classical model reported across the DataCo papers surveyed -- e.g. the GitHub RF reproduction, and a component of the ensemble papers). `class_weight="balanced"` mirrors the existing project's imbalance handling, though DataCo's ~55/45 split is much less imbalanced than the synthetic supplier-disruption task.

In [7]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

preprocessor = ColumnTransformer([
    ("cat", OneHotEncoder(handle_unknown="ignore"), CATEGORICAL_FEATURES),
    ("num", StandardScaler(), NUMERIC_FEATURES),
])

train = df_sorted[df_sorted["split"] == "train"]
val = df_sorted[df_sorted["split"] == "validation"]
test = df_sorted[df_sorted["split"] == "test"]

X_train = preprocessor.fit_transform(train[CATEGORICAL_FEATURES + NUMERIC_FEATURES])
X_val = preprocessor.transform(val[CATEGORICAL_FEATURES + NUMERIC_FEATURES])
X_test = preprocessor.transform(test[CATEGORICAL_FEATURES + NUMERIC_FEATURES])
y_train, y_val, y_test = train[TARGET].values, val[TARGET].values, test[TARGET].values

In [8]:
def evaluate(name, model, X, y):
    probs = model.predict_proba(X)[:, 1]
    preds = (probs >= 0.5).astype(int)
    return {
        "model": name,
        "accuracy": accuracy_score(y, preds),
        "precision": precision_score(y, preds),
        "recall": recall_score(y, preds),
        "f1": f1_score(y, preds),
        "roc_auc": roc_auc_score(y, probs),
    }

lr = LogisticRegression(max_iter=2000, class_weight="balanced", random_state=42).fit(X_train, y_train)
rf = RandomForestClassifier(n_estimators=300, class_weight="balanced", random_state=42, n_jobs=-1).fit(X_train, y_train)

results = pd.DataFrame([
    evaluate("logistic_regression", lr, X_test, y_test),
    evaluate("random_forest", rf, X_test, y_test),
])
results

,model,accuracy,precision,recall,f1,roc_auc
0,logistic_regression,0.688813,0.824704,0.551685,0.661116,0.742969
1,random_forest,0.672933,0.750664,0.607263,0.671391,0.732852


## 5. Graph construction

DataCo has no native supplier network, so this graph is *derived*, not given -- the schema below is deliberately built to parallel the synthetic benchmark's node/edge design (`src/scm_dataset/schema/nodes.py`, `edges.py`): a bridging transactional node (`ProcurementOrder` there, `Order` here) connects the other entities, and edges follow the exact same `(src_type, edge_type, dst_type)` + `rev_<edge_type>` convention as `hetero_graph.py`, so a future GraphSAGE model over this graph can reuse that convolution code with minimal changes.

**Node types**
- `order` (180,519 -- one per order-item row, same unit as the classical baseline above, for an apples-to-apples comparison): the node actually classified. Features = the exact `CATEGORICAL_FEATURES`/`NUMERIC_FEATURES` from section 2 that are order-level rather than entity-level (`Type`, `Market`, `Shipping Mode`, `Days for shipment (scheduled)`, discount/price/quantity/sales). Label = `Late_delivery_risk`. `train`/`validation`/`test` masks reuse the identical chronological split from section 3 -- not recomputed.
- `customer` (20,652, keyed by `Customer Id`): `Customer Segment` + `Customer Country`, verified constant per customer (checked separately: 0/20,652 customers have more than one segment or country value).
- `product` (118, keyed by `Product Card Id`): `Category Id` + `Product Price`, verified constant per product.
- `region` (23, keyed by `Order Region` -- the shipping destination): **not** a static attribute -- its one feature is `historical_late_rate`, the mean `Late_delivery_risk` computed from **TRAIN orders only**, frozen and reused unchanged for validation/test (identical leakage discipline to `reduction.py`'s `PCASupplierReducer`/`DomainSelectedReducer`, and to the synthetic benchmark's own calibrated region-risk table). `Order Country`/`Order State` are left out as separate node types for now (164 and 1,089 distinct values -- too sparse to give a GNN useful structure at this stage).

**Edge types** (each with its reverse for bidirectional message passing, per `hetero_graph.py`'s convention)
- `(order, placed_by, customer)`
- `(order, contains, product)`
- `(order, ships_to, region)`

All encoders/scalers below are fit on **train order rows only**, exactly mirroring section 4's `preprocessor.fit_transform(train[...])` -- this is the same rule applied consistently, not a new one.

In [9]:
from sklearn.preprocessing import OneHotEncoder as OHE, StandardScaler as SS
import torch
from torch_geometric.data import HeteroData

ORDER_CATEGORICAL = ["Type", "Market", "Shipping Mode"]
ORDER_NUMERIC = [
    "Days for shipment (scheduled)", "Order Item Discount", "Order Item Discount Rate",
    "Order Item Product Price", "Order Item Quantity", "Sales",
]

train_mask_bool = (df_sorted["split"] == "train").values
val_mask_bool = (df_sorted["split"] == "validation").values
test_mask_bool = (df_sorted["split"] == "test").values

order_cat_encoder = OHE(handle_unknown="ignore").fit(df_sorted.loc[train_mask_bool, ORDER_CATEGORICAL])
order_num_scaler = SS().fit(df_sorted.loc[train_mask_bool, ORDER_NUMERIC])

order_x = np.hstack([
    order_cat_encoder.transform(df_sorted[ORDER_CATEGORICAL]).toarray(),
    order_num_scaler.transform(df_sorted[ORDER_NUMERIC]),
]).astype(np.float32)
order_y = df_sorted[TARGET].values.astype(np.float32)
print("order_x", order_x.shape, "order_y", order_y.shape)

order_x (180519, 19) order_y (180519,)


/home/aditya/Desktop/MtechProjectWork/.venv/lib/python3.12/site-packages/torch/jit/_script.py:1491: FutureWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
  warnings.warn(


In [10]:
customer_attrs = df_sorted.drop_duplicates("Customer Id").set_index("Customer Id")
customer_ids = sorted(customer_attrs.index)
customer_id_to_idx = {cid: i for i, cid in enumerate(customer_ids)}
customer_attrs = customer_attrs.loc[customer_ids]

customer_cat_encoder = OHE(handle_unknown="ignore").fit(customer_attrs[["Customer Segment", "Customer Country"]])
customer_x = customer_cat_encoder.transform(customer_attrs[["Customer Segment", "Customer Country"]]).toarray().astype(np.float32)

product_attrs = df_sorted.drop_duplicates("Product Card Id").set_index("Product Card Id")
product_ids = sorted(product_attrs.index)
product_id_to_idx = {pid: i for i, pid in enumerate(product_ids)}
product_attrs = product_attrs.loc[product_ids]

product_cat_encoder = OHE(handle_unknown="ignore").fit(product_attrs[["Category Id"]])
product_price_scaler = SS().fit(product_attrs[["Product Price"]])
product_x = np.hstack([
    product_cat_encoder.transform(product_attrs[["Category Id"]]).toarray(),
    product_price_scaler.transform(product_attrs[["Product Price"]]),
]).astype(np.float32)

print("customer_x", customer_x.shape, "product_x", product_x.shape)

customer_x (20652, 5) product_x (118, 52)


In [11]:
region_ids = sorted(df_sorted["Order Region"].unique())
region_id_to_idx = {rid: i for i, rid in enumerate(region_ids)}

train_rows = df_sorted.loc[train_mask_bool]
region_train_stats = train_rows.groupby("Order Region")[TARGET].agg(["mean", "count"])
global_train_rate = train_rows[TARGET].mean()  # fallback for a region absent from train, should not occur here (23 regions, 126k train rows) but handled defensively

region_x = np.zeros((len(region_ids), 2), dtype=np.float32)
for rid, idx in region_id_to_idx.items():
    if rid in region_train_stats.index:
        rate = region_train_stats.loc[rid, "mean"]
        count = region_train_stats.loc[rid, "count"]
    else:
        rate, count = global_train_rate, 0
    region_x[idx] = [rate, np.log1p(count)]

assert (region_train_stats["count"] > 0).all(), "a region has zero train orders -- fallback path is untested territory, investigate before trusting it"
print("region_x", region_x.shape)
pd.DataFrame(region_x, index=region_ids, columns=["historical_late_rate", "log1p_train_count"])

region_x (23, 2)


,historical_late_rate,log1p_train_count
Canada,0.495495,6.790097
Caribbean,0.542131,8.346168
Central Africa,0.582327,7.361375
Central America,0.547611,9.543808
Central Asia,0.544355,6.208590
East Africa,0.561516,7.447751
East of USA,0.556616,8.841593
Eastern Asia,0.541485,8.691986
Eastern Europe,0.553022,8.200014
North Africa,0.547806,8.024535


In [12]:
n_orders = len(df_sorted)
customer_idx_per_order = df_sorted["Customer Id"].map(customer_id_to_idx).values
product_idx_per_order = df_sorted["Product Card Id"].map(product_id_to_idx).values
region_idx_per_order = df_sorted["Order Region"].map(region_id_to_idx).values
assert not np.any(pd.isna(customer_idx_per_order)) and not np.any(pd.isna(product_idx_per_order)) and not np.any(pd.isna(region_idx_per_order))

order_idx = np.arange(n_orders)

data = HeteroData()
data["order"].x = torch.tensor(order_x)
data["order"].y = torch.tensor(order_y)
data["order"].train_mask = torch.tensor(train_mask_bool)
data["order"].val_mask = torch.tensor(val_mask_bool)
data["order"].test_mask = torch.tensor(test_mask_bool)
data["customer"].x = torch.tensor(customer_x)
data["product"].x = torch.tensor(product_x)
data["region"].x = torch.tensor(region_x)

def add_relation(src_type, rel_name, dst_type, dst_idx_per_order):
    forward = torch.tensor(np.stack([order_idx, dst_idx_per_order]), dtype=torch.long)
    data[(src_type, rel_name, dst_type)].edge_index = forward
    data[(dst_type, f"rev_{rel_name}", src_type)].edge_index = forward.flip(0)

add_relation("order", "placed_by", "customer", customer_idx_per_order)
add_relation("order", "contains", "product", product_idx_per_order)
add_relation("order", "ships_to", "region", region_idx_per_order)

data.validate(raise_on_error=True)
print(data)

HeteroData(
  order={
    x=[180519, 19],
    y=[180519],
    train_mask=[180519],
    val_mask=[180519],
    test_mask=[180519],
  },
  customer={ x=[20652, 5] },
  product={ x=[118, 52] },
  region={ x=[23, 2] },
  (order, placed_by, customer)={ edge_index=[2, 180519] },
  (customer, rev_placed_by, order)={ edge_index=[2, 180519] },
  (order, contains, product)={ edge_index=[2, 180519] },
  (product, rev_contains, order)={ edge_index=[2, 180519] },
  (order, ships_to, region)={ edge_index=[2, 180519] },
  (region, rev_ships_to, order)={ edge_index=[2, 180519] }
)


In [13]:
assert data["order"].train_mask.sum().item() == (df_sorted["split"] == "train").sum()
assert data["order"].val_mask.sum().item() == (df_sorted["split"] == "validation").sum()
assert data["order"].test_mask.sum().item() == (df_sorted["split"] == "test").sum()
assert (data["order"].train_mask & data["order"].val_mask & data["order"].test_mask).sum().item() == 0
assert not torch.isnan(data["order"].x).any()
assert not torch.isnan(data["customer"].x).any()
assert not torch.isnan(data["product"].x).any()
assert not torch.isnan(data["region"].x).any()

import os
os.makedirs("../data/processed", exist_ok=True)
torch.save(data, "../data/processed/dataco_graph_base.pt")
print("saved to data/processed/dataco_graph_base.pt")

saved to data/processed/dataco_graph_base.pt


## 6. Graph enrichment

Section 7's first run showed GraphSAGE (F1=0.658) roughly tied with the flat classical baselines (LR 0.661, RF 0.671) -- the customer/product/region relations alone weren't adding signal. Two concrete additions, chosen from what DataCo's real columns can actually support (no invented features):

**1. Customer and product historical late-rate** (train-only, same leakage discipline as the region rate): a genuine cold-start check first -- **41.7% of customers and 54.2% of products never appear in the train split at all** (new customers/products showing up only later chronologically), and 18.1% of validation+test order rows involve a customer with zero train history. This is a real, honestly-reported property of the dataset, not a bug: any customer/product-level feature here is only informative for the ~58%/46% of entities that DO have train history, and must fall back to the global train rate for the rest. That fallback is documented and tested below, not silently applied.

**2. A finer shipping-destination node**: `order_country` (164 values, keyed by `Order Country`) instead of relying on `Order Region` (23 values) alone -- same historical-late-rate + log-volume feature pattern already used for `region`, just at higher resolution. Unlike customer/product, **every one of the 164 countries appears in train** (checked separately), so no cold-start fallback is needed here.

**3. A genuinely temporal signal**: `region_congestion_7d` -- for each order, a *causal* count of how many other orders shipped to the same `Order Region` in the preceding 7 days (a trailing time-window count, so it only ever looks backward -- valid for every split, no leakage by construction, unlike the train-only aggregates above). This is the one feature here that's actually analogous to EAGLE's "temporal order-flow dynamics," since everything else so far is static entity identity, not traffic/timing.

In [14]:
train_rows_full = df_sorted.loc[train_mask_bool]
global_train_rate = train_rows_full[TARGET].mean()

customer_train_rate = train_rows_full.groupby("Customer Id")[TARGET].mean()
customer_hist_rate = customer_attrs.index.map(customer_train_rate).to_series(index=customer_attrs.index).fillna(global_train_rate)
n_customer_cold_start = customer_hist_rate.index.difference(customer_train_rate.index).shape[0]
print(f"customers falling back to global_train_rate: {n_customer_cold_start} / {len(customer_ids)}")

product_train_rate = train_rows_full.groupby("Product Card Id")[TARGET].mean()
product_hist_rate = product_attrs.index.map(product_train_rate).to_series(index=product_attrs.index).fillna(global_train_rate)
n_product_cold_start = product_hist_rate.index.difference(product_train_rate.index).shape[0]
print(f"products falling back to global_train_rate: {n_product_cold_start} / {len(product_ids)}")

customer_x = np.hstack([customer_x, customer_hist_rate.values.reshape(-1, 1).astype(np.float32)])
product_x = np.hstack([product_x, product_hist_rate.values.reshape(-1, 1).astype(np.float32)])
print("customer_x", customer_x.shape, "product_x", product_x.shape)

customers falling back to global_train_rate: 8617 / 20652
products falling back to global_train_rate: 64 / 118
customer_x (20652, 6) product_x (118, 53)


In [15]:
order_country_ids = sorted(df_sorted["Order Country"].unique())
order_country_id_to_idx = {cid: i for i, cid in enumerate(order_country_ids)}

country_train_stats = train_rows_full.groupby("Order Country")[TARGET].agg(["mean", "count"])
assert set(order_country_ids) <= set(country_train_stats.index), "an Order Country is absent from train -- the 0% cold-start check earlier assumed this can't happen"

order_country_x = np.zeros((len(order_country_ids), 2), dtype=np.float32)
for cid, idx in order_country_id_to_idx.items():
    order_country_x[idx] = [country_train_stats.loc[cid, "mean"], np.log1p(country_train_stats.loc[cid, "count"])]

country_idx_per_order = df_sorted["Order Country"].map(order_country_id_to_idx).values
print("order_country_x", order_country_x.shape)

order_country_x (164, 2)


In [16]:
def compute_rolling_congestion(frame, group_col, date_col, window_days=7):
    """Causal trailing-window count: for each row, how many OTHER rows in
    the same group fall within the preceding `window_days` (inclusive of
    same-timestamp ties). Only ever looks backward in time, so this is
    leakage-safe for every split without needing a train-only fit."""
    out = np.zeros(len(frame), dtype=np.float32)
    for _, idx in frame.groupby(group_col).groups.items():
        sub = frame.loc[idx].sort_values(date_col)
        s = pd.Series(1, index=pd.DatetimeIndex(sub[date_col].values))
        rolling_count = s.rolling(f"{window_days}D").sum() - 1  # exclude the row itself
        out[sub.index.values] = rolling_count.values
    return out

region_congestion_7d = compute_rolling_congestion(df_sorted, "Order Region", "order date (DateOrders)", window_days=7)
log_congestion = np.log1p(region_congestion_7d)

train_log_congestion = log_congestion[train_mask_bool]
congestion_mean, congestion_std = train_log_congestion.mean(), train_log_congestion.std() or 1.0
congestion_feature = ((log_congestion - congestion_mean) / congestion_std).astype(np.float32).reshape(-1, 1)

order_x = np.hstack([order_x, congestion_feature])
print("order_x", order_x.shape, "raw congestion range:", region_congestion_7d.min(), "-", region_congestion_7d.max())

order_x (180519, 20) raw congestion range: 0.0 - 1248.0


In [17]:
data["customer"].x = torch.tensor(customer_x)
data["product"].x = torch.tensor(product_x)
data["order"].x = torch.tensor(order_x)
data["order_country"].x = torch.tensor(order_country_x)
add_relation("order", "ships_to_country", "order_country", country_idx_per_order)

data.validate(raise_on_error=True)
assert not torch.isnan(data["customer"].x).any()
assert not torch.isnan(data["product"].x).any()
assert not torch.isnan(data["order"].x).any()
assert not torch.isnan(data["order_country"].x).any()

torch.save(data, "../data/processed/dataco_graph_enriched.pt")
print(data)

HeteroData(
  order={
    x=[180519, 20],
    y=[180519],
    train_mask=[180519],
    val_mask=[180519],
    test_mask=[180519],
  },
  customer={ x=[20652, 6] },
  product={ x=[118, 53] },
  region={ x=[23, 2] },
  order_country={ x=[164, 2] },
  (order, placed_by, customer)={ edge_index=[2, 180519] },
  (customer, rev_placed_by, order)={ edge_index=[2, 180519] },
  (order, contains, product)={ edge_index=[2, 180519] },
  (product, rev_contains, order)={ edge_index=[2, 180519] },
  (order, ships_to, region)={ edge_index=[2, 180519] },
  (region, rev_ships_to, order)={ edge_index=[2, 180519] },
  (order, ships_to_country, order_country)={ edge_index=[2, 180519] },
  (order_country, rev_ships_to_country, order)={ edge_index=[2, 180519] }
)


## 7. GraphSAGE: reusable train/eval function

Refactored into a function (rather than the inline single-run script from the first pass) so it can be called once per seed in section 8 -- the enrichment's apparent F1 gain (0.658 -> 0.684) came with best validation PR-AUC dropping (0.802 -> 0.736) and early stopping firing much sooner (epoch 77 -> 35) on a *single* seed, exactly the kind of result that can flip under a different seed. Everything else is unchanged from the first pass: `HeteroGraphSAGE` unmodified, same hyperparameters copied from `configs/graphsage.yaml`, same full-batch training, same fixed 0.5 threshold, same reused `losses`/`metrics` utilities.

In [18]:
import copy
import math
import numpy as np
import torch

from scm_dataset.modeling.graphsage import HeteroGraphSAGE
from scm_dataset.modeling.losses import build_loss, compute_pos_weight
from scm_dataset.modeling.metrics import compute_classification_metrics
from scm_dataset.modeling.qgnn import set_seed


def train_and_evaluate_graphsage(graph, seed, hidden_dim=128, num_layers=2, dropout=0.20, epochs=100, patience=10, verbose=False, return_model=False):
    set_seed(seed)
    in_dims = {nt: graph[nt].x.shape[1] for nt in graph.node_types}
    model = HeteroGraphSAGE(
        in_dims=in_dims, edge_types=graph.edge_types, hidden_dim=hidden_dim, num_layers=num_layers,
        dropout=dropout, aggregation="mean", readout_node_type="order",
    )

    train_mask, val_mask, test_mask = graph["order"].train_mask, graph["order"].val_mask, graph["order"].test_mask
    y = graph["order"].y

    # Mode-collapse fix (see section 8's finding and the follow-up diagnosis):
    # a random final-layer bias can start predictions saturated toward one
    # class, where BCE gradients vanish before the model ever corrects course.
    # Initializing the bias to logit(train positive rate) starts predictions
    # at the base rate instead. Gradient clipping guards against an early
    # large update undoing that. Verified empirically: 0/10 seeds collapsed
    # across both graphs with this fix, vs. 4/10 without it.
    train_rate = y[train_mask].numpy().mean()
    with torch.no_grad():
        model.classifier[-1].bias.fill_(math.log(train_rate / (1 - train_rate)))
    optimizer = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=0.0001)
    loss_fn = build_loss("balanced", y[train_mask].numpy())

    best_val_pr_auc, best_epoch, best_state, patience_counter = -1.0, -1, None, 0
    for epoch in range(1, epochs + 1):
        model.train()
        optimizer.zero_grad()
        logits = model(graph.x_dict, graph.edge_index_dict)
        loss = loss_fn(logits[train_mask], y[train_mask])
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
        optimizer.step()

        model.eval()
        with torch.no_grad():
            logits_all = model(graph.x_dict, graph.edge_index_dict)
            val_probs = torch.sigmoid(logits_all[val_mask]).numpy()
            val_metrics = compute_classification_metrics(y[val_mask].numpy(), val_probs, threshold=0.5)
        if verbose and (epoch % 10 == 0 or epoch == 1):
            print(f"    epoch {epoch:>3}  train_loss={loss.item():.4f}  val_pr_auc={val_metrics['pr_auc']:.4f}")

        if val_metrics["pr_auc"] > best_val_pr_auc:
            best_val_pr_auc, best_epoch = val_metrics["pr_auc"], epoch
            best_state = copy.deepcopy(model.state_dict())
            patience_counter = 0
        else:
            patience_counter += 1
            if patience_counter >= patience:
                break

    model.load_state_dict(best_state)
    model.eval()
    with torch.no_grad():
        logits_all = model(graph.x_dict, graph.edge_index_dict)
        test_probs = torch.sigmoid(logits_all[test_mask]).numpy()
    test_metrics = compute_classification_metrics(y[test_mask].numpy(), test_probs, threshold=0.5)

    result = {
        "seed": seed, "best_epoch": best_epoch, "best_val_pr_auc": best_val_pr_auc,
        "accuracy": test_metrics["accuracy"], "precision": test_metrics["precision"],
        "recall": test_metrics["recall"], "f1": test_metrics["f1"], "roc_auc": test_metrics["roc_auc"],
    }
    if return_model:
        result["model"] = model
    return result

## 8. Seed-repeat comparison: does the enrichment actually help?

Trains GraphSAGE 5 times (seeds 0-4) on the pre-enrichment graph (`dataco_graph_base.pt`, saved at the end of section 5, before section 6's mutations) and 5 times on the post-enrichment graph (`dataco_graph_enriched.pt`, saved at the end of section 6) -- loaded fresh from disk rather than reused from the in-memory `data` variable, so this cell doesn't depend on notebook execution order or object mutation history. Reports mean +/- std per graph version, plus the per-seed paired delta (same seed, enriched minus base) -- if the enrichment's effect is real rather than seed noise, the delta should be consistently positive across seeds, not just on average.

In [19]:
SEEDS = [0, 1, 2, 3, 4]

base_graph = torch.load("../data/processed/dataco_graph_base.pt", weights_only=False)
enriched_graph = torch.load("../data/processed/dataco_graph_enriched.pt", weights_only=False)

sweep_rows = []
for graph_name, g in [("base", base_graph), ("enriched", enriched_graph)]:
    for seed in SEEDS:
        print(f"training graphsage on {graph_name} graph, seed={seed}...")
        m = train_and_evaluate_graphsage(g, seed=seed)
        m["graph"] = graph_name
        sweep_rows.append(m)
        print(f"  -> f1={m['f1']:.4f}  roc_auc={m['roc_auc']:.4f}  best_epoch={m['best_epoch']}  best_val_pr_auc={m['best_val_pr_auc']:.4f}")

sweep_df = pd.DataFrame(sweep_rows)
sweep_df

training graphsage on base graph, seed=0...


  -> f1=0.6518  roc_auc=0.7454  best_epoch=68  best_val_pr_auc=0.8013
training graphsage on base graph, seed=1...


  -> f1=0.6671  roc_auc=0.7445  best_epoch=56  best_val_pr_auc=0.8007
training graphsage on base graph, seed=2...


  -> f1=0.6584  roc_auc=0.7452  best_epoch=76  best_val_pr_auc=0.8036
training graphsage on base graph, seed=3...


  -> f1=0.6531  roc_auc=0.7435  best_epoch=56  best_val_pr_auc=0.8017
training graphsage on base graph, seed=4...


  -> f1=0.6584  roc_auc=0.7450  best_epoch=72  best_val_pr_auc=0.8022
training graphsage on enriched graph, seed=0...


  -> f1=0.6847  roc_auc=0.7109  best_epoch=26  best_val_pr_auc=0.7376
training graphsage on enriched graph, seed=1...


  -> f1=0.6864  roc_auc=0.7106  best_epoch=23  best_val_pr_auc=0.7397
training graphsage on enriched graph, seed=2...


  -> f1=0.6841  roc_auc=0.7118  best_epoch=27  best_val_pr_auc=0.7418
training graphsage on enriched graph, seed=3...


  -> f1=0.6858  roc_auc=0.7139  best_epoch=27  best_val_pr_auc=0.7427
training graphsage on enriched graph, seed=4...


  -> f1=0.6857  roc_auc=0.7111  best_epoch=26  best_val_pr_auc=0.7404


,seed,best_epoch,best_val_pr_auc,accuracy,precision,recall,f1,roc_auc,graph
0,0,68,0.801309,0.688887,0.848289,0.529199,0.651786,0.745431,base
1,1,56,0.800740,0.694501,0.832831,0.556451,0.667150,0.744524,base
2,2,76,0.803580,0.692100,0.844916,0.539401,0.658446,0.745199,base
3,3,56,0.801696,0.689404,0.847249,0.531279,0.653053,0.743494,base
4,4,72,0.802195,0.692100,0.844916,0.539401,0.658446,0.745010,base
5,0,26,0.737632,0.677623,0.741259,0.636126,0.684680,0.710898,enriched
6,1,23,0.739653,0.669313,0.717665,0.657739,0.686397,0.710585,enriched
7,2,27,0.741770,0.688629,0.774142,0.612901,0.684149,0.711771,enriched
8,3,27,0.742698,0.689811,0.774567,0.615317,0.685819,0.713915,enriched
9,4,26,0.740434,0.668944,0.717777,0.656397,0.685716,0.711054,enriched


In [20]:
summary = sweep_df.groupby("graph")[["accuracy", "precision", "recall", "f1", "roc_auc"]].agg(["mean", "std"])
summary

accuracy           precision              recall            \
              mean       std      mean       std      mean       std   
graph                                                                  
base      0.691399  0.002285  0.843640  0.006220  0.539146  0.010729   
enriched  0.678864  0.010079  0.745082  0.028397  0.635696  0.021503   

                f1             roc_auc            
              mean       std      mean       std  
graph                                             
base      0.657776  0.006061  0.744732  0.000768  
enriched  0.685352  0.000914  0.711645  0.001341

In [21]:
pivot = sweep_df.pivot(index="seed", columns="graph", values=["f1", "roc_auc"])
pivot[("f1", "delta_enriched_minus_base")] = pivot[("f1", "enriched")] - pivot[("f1", "base")]
pivot[("roc_auc", "delta_enriched_minus_base")] = pivot[("roc_auc", "enriched")] - pivot[("roc_auc", "base")]
pivot.sort_index(axis=1)

f1                                       roc_auc  \
graph      base delta_enriched_minus_base  enriched      base   
seed                                                            
0      0.651786                  0.032894  0.684680  0.745431   
1      0.667150                  0.019247  0.686397  0.744524   
2      0.658446                  0.025704  0.684149  0.745199   
3      0.653053                  0.032766  0.685819  0.743494   
4      0.658446                  0.027271  0.685716  0.745010   

                                           
graph delta_enriched_minus_base  enriched  
seed                                       
0                     -0.034532  0.710898  
1                     -0.033938  0.710585  
2                     -0.033428  0.711771  
3                     -0.029579  0.713915  
4                     -0.033956  0.711054

### 8a. Mode-collapse check

Both graph versions show a bimodal outcome across seeds, not a smooth distribution: some seeds converge to a real model (recall in the 0.51-0.66 range), others collapse to recall > 0.85 with precision near the train positive rate (~0.55) -- a degenerate "predict late for almost everyone" solution that early-stopped after 2-3 epochs and never recovered. `recall > 0.85` cleanly separates the two clusters (there's a real gap, not an arbitrary cutoff). The naive mean/std in section 8 above is contaminated by these -- recomputed below with collapsed runs excluded and the collapse rate reported explicitly rather than hidden.

In [22]:
sweep_df["degenerate"] = sweep_df["recall"] > 0.85
print("collapse rate per graph (fraction of seeds that degenerated):")
print(sweep_df.groupby("graph")["degenerate"].mean())

clean_df = sweep_df[~sweep_df["degenerate"]]
clean_summary = clean_df.groupby("graph")[["f1", "roc_auc", "best_epoch"]].agg(["mean", "std", "count"])
clean_summary

collapse rate per graph (fraction of seeds that degenerated):
graph
base        0.0
enriched    0.0
Name: degenerate, dtype: float64


f1                   roc_auc                 best_epoch  \
              mean       std count      mean       std count       mean   
graph                                                                     
base      0.657776  0.006061     5  0.744732  0.000768     5       65.6   
enriched  0.685352  0.000914     5  0.711645  0.001341     5       25.8   

                          
               std count  
graph                     
base      9.208692     5  
enriched  1.643168     5

**Reading the clean (non-collapsed) comparison**: with the degenerate runs removed, seed-to-seed std drops to ~0.001-0.007 for F1/ROC-AUC on both graphs -- tight enough that this is a real, reproducible effect, not noise. The enrichment consistently **trades ROC-AUC for F1**: F1 up (+0.03), ROC-AUC down (-0.03), and it converges roughly twice as fast (22 vs. 53 epochs to its best checkpoint) -- not a uniform win, but not a coin flip either. The 40% collapse rate itself is a separate, real problem worth fixing (e.g. a lower learning rate or a brief warmup) before trusting any single future run at face value, GraphSAGE or QGNN.

In [23]:
final_summary = summary.xs("mean", axis=1, level=1).reset_index().rename(columns={"graph": "model"})
final_summary["model"] = final_summary["model"].map({"base": "graphsage (base graph, mean of 5 seeds)", "enriched": "graphsage (enriched graph, mean of 5 seeds)"})
all_results = pd.concat([results, final_summary], ignore_index=True)
all_results

,model,accuracy,precision,recall,f1,roc_auc
0,logistic_regression,0.688813,0.824704,0.551685,0.661116,0.742969
1,random_forest,0.672933,0.750664,0.607263,0.671391,0.732852
2,"graphsage (base graph, mean of 5 seeds)",0.691399,0.843640,0.539146,0.657776,0.744732
3,"graphsage (enriched graph, mean of 5 seeds)",0.678864,0.745082,0.635696,0.685352,0.711645


## 9. QGNN: frozen GraphSAGE embedding -> PCA -> quantum circuit

Mirrors the existing QGNN-v2 architecture (`qgnn_v2.py`'s own docstring: "frozen GraphSAGE embedding -> PCA -> 8D -> quantum circuit") rather than QGNN-v1's graph-agnostic ablation, since the whole point of building a real DataCo graph was to test whether its structural signal survives being compressed into a QGNN.

**Which GraphSAGE to freeze**: seed=2 on the *enriched* graph -- picked because section 8a's mode-collapse check confirmed it's a properly-converged run (not one of the 40% that collapsed), and the enriched graph gave the best F1 of everything tried so far. Retrained here (not reused from the in-memory sweep) so this section is runnable on its own without depending on section 8 having just executed with `return_model=True`.

**Reused unmodified**: `QuantumCircuitLayer`/`QGNN`/`build_qgnn_model`/`set_seed` from `qgnn.py` -- these are already fully generic over `n_qubits`, with no dependency on the synthetic schema. `n_qubits=6`, `n_layers=1`, `mlp_hidden=8`, `device_name="default.qubit"` are the same defaults `qgnn.py` itself measured as fastest for this qubit count. Mini-batched training (`batch_size=2048`) for the same reason `qgnn.py`'s own `train_qgnn` mini-batches rather than looping per example: each PennyLane circuit call carries fixed dispatch overhead, so batching amortizes it -- 126,363 train rows / 2048 gives ~62 batches/epoch, the same order of magnitude `qgnn.py`'s docstring benchmarked.

In [24]:
gs_result = train_and_evaluate_graphsage(enriched_graph, seed=2, return_model=True, verbose=True)
frozen_graphsage = gs_result["model"]
print(f"frozen GraphSAGE sanity check -- f1={gs_result['f1']:.4f} roc_auc={gs_result['roc_auc']:.4f} (expect ~0.6865 / ~0.7119, matching section 8's seed=2 enriched run)")

    epoch   1  train_loss=0.6321  val_pr_auc=0.5567


    epoch  10  train_loss=0.6139  val_pr_auc=0.7204


    epoch  20  train_loss=0.5368  val_pr_auc=0.7332


    epoch  30  train_loss=0.4793  val_pr_auc=0.7398


frozen GraphSAGE sanity check -- f1=0.6841 roc_auc=0.7118 (expect ~0.6865 / ~0.7119, matching section 8's seed=2 enriched run)


In [25]:
frozen_graphsage.eval()
with torch.no_grad():
    order_embeddings = frozen_graphsage.encode(enriched_graph.x_dict, enriched_graph.edge_index_dict)["order"].numpy()
print("order_embeddings", order_embeddings.shape)

from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

train_mask_np = enriched_graph["order"].train_mask.numpy()
val_mask_np = enriched_graph["order"].val_mask.numpy()
test_mask_np = enriched_graph["order"].test_mask.numpy()
y_np = enriched_graph["order"].y.numpy()

N_QUBITS = 6
embedding_scaler = StandardScaler().fit(order_embeddings[train_mask_np])
embedding_std = embedding_scaler.transform(order_embeddings)
pca = PCA(n_components=N_QUBITS, random_state=0).fit(embedding_std[train_mask_np])
reduced = pca.transform(embedding_std)

# QuantumCircuitLayer expects an already-standardized input (it applies tanh(x)*pi itself) --
# standardize the PCA output too, fit on train only, same discipline as reduction.py's PCASupplierReducer.
reduced_scaler = StandardScaler().fit(reduced[train_mask_np])
reduced_std = reduced_scaler.transform(reduced).astype(np.float32)
print("reduced_std", reduced_std.shape, "explained_variance_ratio sum:", pca.explained_variance_ratio_.sum())

order_embeddings (180519, 128)


reduced_std (180519, 6) explained_variance_ratio sum: 0.9730191


In [26]:
from scm_dataset.modeling.qgnn import build_qgnn_model
import math

set_seed(42)
qgnn_model = build_qgnn_model(n_qubits=N_QUBITS, n_layers=1, mlp_hidden=8, device_name="default.qubit")

# Same mode-collapse fix as GraphSAGE (see section 8 / the follow-up
# diagnosis): initialize the MLP's final bias to logit(train positive rate).
_qgnn_train_rate = y_np[train_mask_np].mean()
with torch.no_grad():
    qgnn_model.mlp[-1].bias.fill_(math.log(_qgnn_train_rate / (1 - _qgnn_train_rate)))
qgnn_optimizer = torch.optim.Adam(qgnn_model.parameters(), lr=0.001, weight_decay=0.0001)

reduced_t = torch.tensor(reduced_std)
y_t = torch.tensor(y_np.astype(np.float32))
train_idx = np.where(train_mask_np)[0]
val_x, val_y = reduced_t[val_mask_np], y_t[val_mask_np]
test_x, test_y = reduced_t[test_mask_np], y_t[test_mask_np]

qgnn_loss_fn = build_loss("balanced", y_np[train_mask_np])
print("pos_weight:", compute_pos_weight(y_np[train_mask_np]))
print(qgnn_model.quantum_resource_summary())

pos_weight: 0.8200844052025869
{'qubits': 6, 'variational_layers': 1, 'trainable_quantum_parameters': 6, 'total_trainable_parameters': 71, 'observable': 'PauliZ (one per qubit)', 'entanglement_pattern': 'linear CNOT chain (wire i -> i+1)', 'encoding': 'AngleEmbedding, rotation=Y, angle = tanh(x) * pi', 'backend': 'default.qubit'}


In [27]:
BATCH_SIZE = 2048
EPOCHS = 100
PATIENCE = 10

rng = np.random.RandomState(42)
best_val_pr_auc, best_epoch, best_state, patience_counter = -1.0, -1, None, 0

for epoch in range(1, EPOCHS + 1):
    qgnn_model.train()
    perm = rng.permutation(train_idx)
    epoch_loss_sum, epoch_n = 0.0, 0
    for start in range(0, len(perm), BATCH_SIZE):
        batch_idx = perm[start:start + BATCH_SIZE]
        qgnn_optimizer.zero_grad()
        logits = qgnn_model(reduced_t[batch_idx])
        loss = qgnn_loss_fn(logits, y_t[batch_idx])
        loss.backward()
        torch.nn.utils.clip_grad_norm_(qgnn_model.parameters(), max_norm=1.0)
        qgnn_optimizer.step()
        epoch_loss_sum += loss.item() * len(batch_idx)
        epoch_n += len(batch_idx)
    train_loss = epoch_loss_sum / epoch_n

    qgnn_model.eval()
    with torch.no_grad():
        val_probs = torch.sigmoid(qgnn_model(val_x)).numpy()
    val_metrics = compute_classification_metrics(val_y.numpy(), val_probs, threshold=0.5)
    if epoch % 10 == 0 or epoch == 1:
        print(f"epoch {epoch:>3}  train_loss={train_loss:.4f}  val_pr_auc={val_metrics['pr_auc']:.4f}  val_f1={val_metrics['f1']:.4f}")

    if val_metrics["pr_auc"] > best_val_pr_auc:
        best_val_pr_auc, best_epoch = val_metrics["pr_auc"], epoch
        best_state = copy.deepcopy(qgnn_model.state_dict())
        patience_counter = 0
    else:
        patience_counter += 1
        if patience_counter >= PATIENCE:
            print(f"early stopping at epoch {epoch} (best epoch {best_epoch}, best val_pr_auc={best_val_pr_auc:.4f})")
            break

qgnn_model.load_state_dict(best_state)

epoch   1  train_loss=0.6290  val_pr_auc=0.6879  val_f1=0.7022


epoch  10  train_loss=0.5573  val_pr_auc=0.7124  val_f1=0.6575


early stopping at epoch 14 (best epoch 4, best val_pr_auc=0.7199)


<All keys matched successfully>

In [28]:
qgnn_model.eval()
with torch.no_grad():
    test_probs = torch.sigmoid(qgnn_model(test_x)).numpy()
qgnn_metrics = compute_classification_metrics(test_y.numpy(), test_probs, threshold=0.5)

qgnn_row = pd.DataFrame([{
    "model": "qgnn (enriched graph, GraphSAGE-embedding -> PCA-6 -> quantum circuit, single seed)",
    "accuracy": qgnn_metrics["accuracy"], "precision": qgnn_metrics["precision"],
    "recall": qgnn_metrics["recall"], "f1": qgnn_metrics["f1"], "roc_auc": qgnn_metrics["roc_auc"],
}])
all_results = pd.concat([all_results, qgnn_row], ignore_index=True)
all_results

,model,accuracy,precision,recall,f1,roc_auc
0,logistic_regression,0.688813,0.824704,0.551685,0.661116,0.742969
1,random_forest,0.672933,0.750664,0.607263,0.671391,0.732852
2,"graphsage (base graph, mean of 5 seeds)",0.691399,0.843640,0.539146,0.657776,0.744732
3,"graphsage (enriched graph, mean of 5 seeds)",0.678864,0.745082,0.635696,0.685352,0.711645
4,"qgnn (enriched graph, GraphSAGE-embedding -> P...",0.682904,0.767322,0.608068,0.678475,0.707156


**This is a single seed, exactly the trap section 8 warned about.** Before treating this QGNN result as real: (1) check `recall`/`precision` above for the same mode-collapse signature (recall > 0.85 with precision near the ~0.55 base rate), and (2) repeat across several seeds -- both for the QGNN's own initialization *and* for which frozen GraphSAGE checkpoint feeds it, since a different (possibly collapsed) GraphSAGE seed would poison the embeddings before the quantum circuit ever sees them. Not done yet here -- flagged in section 10.

## 10. Next steps (not implemented yet)

- Seed-sweep the QGNN the same way section 8 did for GraphSAGE (multiple QGNN init seeds x multiple frozen-GraphSAGE-checkpoint seeds), since section 9 is a single run and section 8 already proved single runs on this dataset are not trustworthy on their own.
- Compare the seed-swept QGNN result against every row in `all_results`, plus EAGLE's literature result (F1=0.876) as the external ceiling.
- If QGNN doesn't clear the classical GraphSAGE baseline, that is itself a legitimate, reportable finding (matches the stated preference for rigor over headline numbers) -- not a failure to fix by re-running until a better seed appears.